pkg/sentry/fsimpl/fuse/connection.go

```diff
@@ -0,0 +1,208 @@
+}
+
+// NewFUSEConnection creates a FUSE connection to fd
+func NewFUSEConnection(ctx context.Context, fd *vfs.FileDescription) *Connection {
+	// Mark the device as ready so it can be used. /dev/fuse can only be used if the FD was used to
+	// mount a FUSE filesystem.
+	fuseFD := fd.Impl().(*DeviceFD)
+	fuseFD.mounted = true
+
+	// Create the writeBuf for the header to be stored in.
+	hdrLen := uint32((*linux.FUSEHeaderOut)(nil).SizeBytes())
+	fuseFD.writeBuf = make([]byte, hdrLen)
+	fuseFD.completions = make(map[linux.FUSEOpID]*FutureResponse)
+	fuseFD.waitCh = make(chan struct{}, MaxInFlightRequests)
+	fuseFD.writeCursor = 0
+	fuseFD.readCursor = 0
+
+	return &Connection{fd: fuseFD}
+}
+
+// NewRequest creates a new request that can be sent to the FUSE server.
+func (conn *Connection) NewRequest(creds *auth.Credentials, pid uint32, ino uint64, opcode linux.FUSEOpcode, payload marshal.Marshallable) (*Request, error) {
+	conn.fd.mu.Lock()
+	defer conn.fd.mu.Unlock()
+
+	hdrLen := (*linux.FUSEHeaderIn)(nil).SizeBytes()
+	hdr := linux.FUSEHeaderIn{
+		Len:    uint32(hdrLen + payload.SizeBytes()),
+		Opcode: opcode,
+		Unique: conn.fd.nextOpID,
+		NodeID: ino,
+		UID:    uint32(creds.EffectiveKUID),
+		GID:    uint32(creds.EffectiveKGID),
+		PID:    pid,
+	}
+	conn.fd.nextOpID++
+
+	buf := make([]byte, hdr.Len)
+	hdr.MarshalUnsafe(buf[:hdrLen])
+	payload.MarshalUnsafe(buf[hdrLen:])
+
+	return &Request{
+		id:   hdr.Unique,
+		hdr:  &hdr,
+		data: buf,
+	}, nil
+}
+
+// Call makes a request to the server and blocks the invoking task until a
+// server responds with a response.
+func (conn *Connection) Call(t *kernel.Task, r *Request) (*Response, error) {
+	fut, err := conn.callFuture(r)
+	if err != nil {
+		return nil, err
+	}
+
+	return fut.resolve(t)
+}
+
+// Call makes a request to the server and blocks the invoking goroutine until a
+// server responds with a response. Doesn't block a kernel.Task.
+func (conn *Connection) CallTaskNonBlock(r *Request) (*Response, error) {
+	fut, err := conn.callFuture(r)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: Consider allowing a timeout to unblock the process. Unclear if this
+	// is required yet.
+	select {
+	case <-fut.ch:
+	}
+
+	// A response is ready. Resolve and return it.
+	return &Response{
+		hdr:  *fut.hdr,
+		data: fut.data,
+	}, nil
+}
+
+// callFuture makes a request to the server and returns a future response.
+// Call resolve() when the response needs to be fulfilled.
+func (conn *Connection) callFuture(r *Request) (*FutureResponse, error) {
+	conn.fd.mu.Lock()
+	conn.fd.queue.PushBack(r)
+	fut := newFutureResponse()
+	conn.fd.completions[r.id] = fut
+	conn.fd.mu.Unlock()
+
+	// Signal a reader notifying them about a queued request. This
+	// might block if the number of in flight requests exceed
+	// MaxInFlightRequests.
+	//
+	// TODO: Consider possible starvation here if the waitCh is
+	// continuously full. Will go channels respect FIFO order when
+	// unblocking threads?
```

Comment:
@mrahatm @jinmouil @avagin Right now this will just simply block if the waitCh is full. See the TODO. How do we (if at all) deal with this? Can this be punted until a bit down the road - I can create an issue for it.

Reply:
Is this analogous? https://elixir.bootlin.com/linux/latest/source/fs/fuse/dev.c#L298